File: 1046-last-stone-weight/1046-last-stone-weight.java
class Solution {
    public int lastStoneWeight(int[] stones) {
        PriorityQueue<Integer> pq = new PriorityQueue<>((x, y) -> (y - x));
        
        for(int stone : stones){
            pq.offer(stone);
        }
        
        while(pq.size() > 1){
            int s1 = pq.poll();
            int s2 = pq.poll();
            if(s1 != s2){
                pq.offer(Math.abs(s1 - s2));
            }
        }
        
        return pq.isEmpty() ? 0 : pq.peek();
    }
}

File: 1046-last-stone-weight/1046-last-stone-weight.py
import heapq
class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        max_heap = []
        for s in stones:
            heapq.heappush(max_heap, -s)
        
        while len(max_heap) > 1:
            s1 = -heapq.heappop(max_heap)
            s2 = -heapq.heappop(max_heap)
            if s1 != s2:
                heapq.heappush(max_heap, s2 - s1)
        
        return -max_heap[0] if max_heap else 0